File: my-folder/0207-course-schedule/solution.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegree = [0] * numCourses
        adj_list = defaultdict(list)
        for crs, pre in prerequisites:
            adj_list[pre].append(crs)
            indegree[crs] += 1
        
        queue = deque([crs for crs in range(numCourses) if indegree[crs] == 0])
        
        visited = 0
        while queue:
            curr_course = queue.popleft()
            visited += 1
            for nei in adj_list[curr_course]:
                indegree[nei] -= 1
                if(indegree[nei] == 0):
                    queue.append(nei)
        
        if visited == numCourses:
            return True
        return False

File: my-folder/0207-course-schedule/solution.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adjList = new ArrayList<>();
        int[] indegree = new int[numCourses];

        for(int i = 0; i < numCourses; i++){
            adjList.add(new ArrayList<Integer>());
        }   

        for(int[] prereq : prerequisites){
            int crs = prereq[0];
            int pre = prereq[1];
            indegree[crs]++;
            adjList.get(pre).add(crs);
        }

        int finish = 0;
        Queue<Integer> completed = new LinkedList<>();
        for(int i = 0; i < numCourses; i++){
            if(indegree[i] == 0){
                completed.offer(i);
            }
        }

        while(!completed.isEmpty()){
            int curr = completed.poll();
            finish++;
            for(int nei : adjList.get(curr)){
                indegree[nei]--;
                if(indegree[nei] == 0){
                    completed.offer(nei);
                }
            }
        }

        return finish == numCourses;
    }
}




